My courses

emilkovacev opened the pull request:
## Proposed changes
--------------------
closes #15 
- Adds a user interface and drop-down filter for the Courses page
  - Displays a list of active courses
  - Saves filter option in local storage
  - Adds global styles

## Types of changes
--------------------
What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist
--------------------
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [x] My changeset covers only what is described above (no extraneous changes)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged downstream

## Further comments

I added a small tag to the left of every course to easily differentiate them and add some color. I used an [npm package](https://github.com/zenozeng/color-hash) licensed under MIT license to make hashes of each course code. I think it adds a nice visual flair, but if it needs to be removed then that's completely fine.

I didn't add this to this pull request, but there's an argument to be made to be able to toggle between rows and cards (which given the current implementation would not be difficult to do). Possibly redundant but thought I would at least throw the idea out there.

--------------------
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

![devU-courses](https://user-images.githubusercontent.com/56481968/127257037-e9fcc528-35ba-4f75-9f1d-0e7d82d6e1ff.gif)

![devU-client-filter](https://user-images.githubusercontent.com/56481968/127257745-e6a3b51a-20be-4ef7-a30c-6b0598bb4353.gif)

![devU-client-resize](https://user-images.githubusercontent.com/56481968/127335803-d0130b86-2ed1-4380-aa69-7c6d290b092a.gif)

emilkovacev committed:
* Initial Changes

* added CourseView, CourseFilter (in-progress), and CourseList (list of CourseViews)
* added filter dropdown
* reformatted information to table view
* transitioned grid to using flexbox for versatility

MikeDrewitt committed:
* Fetching data from API
* Displaying all user-course & course info

MikeDrewitt committed:
Cleaned up CSS

MikeDrewitt committed:
Color updates

MikeDrewitt committed:
* Added dropdown as filter

* TODO
  * Store filter changes into local storage and read from ls for default
  * Touchup css (make it prettier)
  * Use actual loading overlay component
  * Use actual error component

emilkovacev committed:
added custom styles for listitems

emilkovacev committed:
stylized list components

emilkovacev committed:
Adds functionality to set filter value in localhost
Modified styling
Added a tag for each ListItem

emilkovacev committed:
Added tag-color generated by hash; Added header for course page

MikeDrewitt commented on src/components/listItems/userCourseListItem.tsx:
```diff
@@ -0,0 +1,38 @@
+  course: Course
+}
+
+const colorHash = (input: string) => {
```
Ya know, my first thought was where you were getting the colors from. This is classy as hell. Really cool idea. 

This is fine for now. But it's very likely that we'll want to pull this out into it's own little utils function so that we can have the same color being used for the course where ever we want rather than just here.

Again this is totally fine for now though. No need to unnecessarily abstract until we need it.

emilkovacev replied:
I assigned myself issue #14 as well, so once we get a working design I can add it as a utils function for that pull request

MikeDrewitt commented on src/components/pages/userCoursesListPage.tsx:
```diff
@@ -1,7 +1,79 @@
+    fetchData()
+  }, [filter])
+
+  if (loading) return <div>Loading</div>
```
Swap this guy for the `<LoadingOverlay />` component

MikeDrewitt commented on src/components/pages/userCoursesListPage.tsx:
```diff
@@ -1,7 +1,79 @@
+  }, [filter])
+
+  if (loading) return <div>Loading</div>
+  if (error) return <div>{error}</div>
```
This is a bit out of scope of what you've done here, but I think it'd be good to add:

We should make an error page that can very cleanly display these errors. You'll find a pretty good example with the `<ValidationErrorViewer />` component. The difference here is make it a full page. So wrap it with it's own `<PageWrapper />` and have it take an error object and display it just like that validation error viewer. This way we can reuse it as an error page basically everywhere.

emilkovacev replied:
Something Like this? I made a commit with my most recent changes.

![devu-error](https://user-images.githubusercontent.com/56481968/127408315-8f3b23a1-0fa3-4928-b668-80760c9b9ecb.gif)

MikeDrewitt replied:
If you're providing the name then sure. We can work on it. But it's good enough. We don't have a name though at least from the error response. So you'd need to provide a title then display the message that comes back with the response

emilkovacev replied:
In that case, I'll remove the Error Name section and leave the error message at the bottom

MikeDrewitt reviewed (changes requested):
Great first pass. The most amount of changes would just be that error page.

Absolutely great work.

emilkovacev committed:
Added full page error-page

Fixed accidental deletion of font

fixed pacing error for error page

Fixed error bug

MikeDrewitt commented on src/components/pages/userCoursesListPage.tsx:
```diff
@@ -1,7 +1,99 @@
+    error: Error | null
+  }
+
+  const ErrorPage = ({error}: Props) => {
```
Oh I'm sorry I wasn't clear. Just make this it's own file. Include the page wrapper in the return statement here. And just call the file errorPage and toss it in the pages directory. 

Also, bonus: you can make this function without an  explicit return to save the nesting

```javascript
const ErrorPage = ({ error }: Props => (
  <PageWrapper>
    // content here
  </PageWrapper>
)
```

emilkovacev replied:
Should we consider placing it in the `src > shared > errors` folder, or is it too circumstantial? Just double-checking to be sure.

MikeDrewitt replied:
Because it's the full page I'd probably put it in pages.

It's possible that we should make an error component that is then called by the errorpage component, but that feels like a bit much for this. If we find we want to display error mesasges elsewhere, we can think about that then

MikeDrewitt commented on src/components/pages/userCoursesListPage.tsx:
```diff
@@ -1,7 +1,99 @@
+  if (loading) return <LoadingOverlay />
+
+  type Props = {
+    error: Error | null
```
don't allow this to be null. No reason to since we should always be passing it an error object.

emilkovacev replied:
How should I fix this? "error" has an initial state of null, it seems like TS doesn't know for certain that it won't be null (even though it's filtered out by the conditional it still marks it as being a problem)

MikeDrewitt replied:
Hmmmmm so the way to fix it would be when returning the error component check if error is truthy before passing it in like this

```javascript
if (error && error !== null) <ErrorPage />
```

But maybe what you have is better cause I don't want to have to do that everywhere. Here's the only thing though. Make sure you check if error is truthy in the error page component though casue otherwise someone could attempt to do `error.message` and get an object undefined. So just have some default error message like "And unknown error has occured" or something. And use error.message if error is not null

emilkovacev replied:
I just made a commit to solve this issue (In errorPage.tsx) and tested it in Chromium, was that the implementation you were looking for?

emilkovacev committed:
Moved ErrorPage into new file in pages repo

MikeDrewitt commented on src/components/pages/userCoursesListPage.tsx:
```diff
@@ -1,7 +1,82 @@
+        <div className={styles.filters}>
+          <Dropdown
+            label='Filter Courses'
+            className={styles.dropdown}
```
This is a super nitpick. There's a small screen size that `My Courses` wrap. I know you're using the screen sizes from variables, but can we make that not use a variable and use a number that forces that to wrap before `My Courses` does.

MikeDrewitt commented on src/components/pages/errorPage.tsx:
```diff
@@ -0,0 +1,21 @@
+        <div className={styles.errorBackground}>
+            <div className={styles.errorContainer}>
+                <h1 className={styles.error}>Error</h1>
+                <p className={styles.errorMessage}>{error?.message}</p>
```
Ah perfect. This is what I was talking about. in this like do 
```suggestion
                <p className={styles.errorMessage}>{error?.message || defaultErrorMessage}</p>
```

then add a `defaultErrorMessage` at the top of this component

emilkovacev committed:
Fixed responsive design inteaction between title and dropdown

emilkovacev committed:
Added default error message if error is null

MikeDrewitt commented:
I'll leave this up until tomorrow if anyone else wants to take a look. But otherwise I think this looks g2g.
